rust/src/domains/polars/series/mod.rs

```diff
@@ -0,0 +1,307 @@
+}
+
+impl Debug for SeriesDomain {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "SeriesDomain(\"{}\", {})",
+            self.field.name, self.field.dtype
+        )
+    }
+}
+
+impl<D: UnboundedMetric> MetricSpace for (SeriesDomain, D) {
+    fn check_space(&self) -> Fallible<()> {
+        Ok(())
+    }
+}
+
+// BEGIN UTILITY TRAITS
+
+/// Common trait for domains that can be used to describe the space of typed elements within a series.
+pub trait SeriesAtomDomain: Domain + Send + Sync {
+    type Atom: CheckAtom + PrimitiveDataType;
+    fn atom_domain(self) -> AtomDomain<Self::Atom>;
+    const NULLABLE: bool;
+}
+impl<T: CheckAtom + PrimitiveDataType> SeriesAtomDomain for AtomDomain<T> {
+    type Atom = T;
+
+    fn atom_domain(self) -> AtomDomain<Self::Atom> {
+        self
+    }
+
+    const NULLABLE: bool = false;
+}
+impl<T: CheckAtom + PrimitiveDataType> SeriesAtomDomain for OptionDomain<AtomDomain<T>> {
+    type Atom = T;
+
+    fn atom_domain(self) -> AtomDomain<Self::Atom> {
+        self.element_domain
+    }
+
+    const NULLABLE: bool = true;
+}
+
+/// Object-safe version of SeriesAtomDomain.
+pub trait DynSeriesAtomDomain: Send + Sync {
+    fn as_any(&self) -> &dyn Any;
+    fn dyn_partial_eq(&self, other: &dyn DynSeriesAtomDomain) -> bool;
+}
+impl<D: 'static + SeriesAtomDomain> DynSeriesAtomDomain for D {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn dyn_partial_eq(&self, other: &dyn DynSeriesAtomDomain) -> bool {
+        (other.as_any().downcast_ref::<D>()).map_or(false, |a| self == a)
+    }
+}
+
+impl PartialEq for dyn DynSeriesAtomDomain + '_ {
+    fn eq(&self, other: &Self) -> bool {
+        self.dyn_partial_eq(other)
+    }
+}
+
+/// Utility trait to construct the Polars runtime data-type indicator from an atomic type.
+pub trait NumericDataType:
+    NumericNative<PolarsType = Self::NumericPolars> + PrimitiveDataType
+{
+    type NumericPolars: PolarsDataType + PolarsNumericType<Native = Self>;
+}
+
+pub trait PrimitiveDataType: 'static + Send + Sync {
+    type Polars: PolarsDataType;
+    fn dtype() -> DataType {
+        Self::Polars::get_dtype()
+    }
+}
+
+macro_rules! impl_dtype_from {
+    ($ty:ty, $dt:ty) => {
+        impl NumericDataType for $ty {
+            type NumericPolars = $dt;
+        }
+        impl PrimitiveDataType for $ty {
+            type Polars = $dt;
+        }
+    };
+}
+impl_dtype_from!(u8, UInt8Type);
+impl_dtype_from!(u16, UInt16Type);
+impl_dtype_from!(u32, UInt32Type);
+impl_dtype_from!(u64, UInt64Type);
+impl_dtype_from!(i8, Int8Type);
+impl_dtype_from!(i16, Int16Type);
+impl_dtype_from!(i32, Int32Type);
+impl_dtype_from!(i64, Int64Type);
+impl_dtype_from!(f32, Float32Type);
+impl_dtype_from!(f64, Float64Type);
+impl PrimitiveDataType for bool {
+    type Polars = BooleanType;
+}
+impl PrimitiveDataType for String {
+    type Polars = StringType;
+}
+#[cfg(test)]
+mod test_series {
```

Comment:
It would be really good to get Rust test coverage working. I can see names up above that aren't referenced here... but maybe they are used indirectly? It's hard for me to know.